lib/mpl_toolkits/mplot3d/axes3d.py

```diff
@@ -412,24 +413,27 @@ def do_3d_projection(artist):
                 Call `do_3d_projection` on an *artist*, and warn if passing
                 *renderer*.
 
-                For our Artists, never pass *renderer*. For external Artists,
-                in lieu of more complicated signature parsing, always pass
-                *renderer* and raise a warning.
+                Attempt to bind the empty signature first, so external Artists
+                can avoid the deprecation warning if they support the new
+                calling convention.
                 """
-
-                if artist.__module__ == 'mpl_toolkits.mplot3d.art3d':
-                    # Our 3D Artists have deprecated the renderer parameter, so
-                    # avoid passing it to them; call this directly once the
-                    # deprecation has expired.
+                try:
+                    signature = inspect.signature(artist.do_3d_projection)
+                    signature.bind()
+                # ValueError if `inspect.signature` cannot provide a signature
+                # and TypeError if the binding fails or the object does not
+                # appear to be callable - the next call will then re-raise.
+                except (ValueError, TypeError):
+                    _api.warn_deprecated(
+                        "3.4",
+                        message="The 'renderer' parameter of "
+                        "do_3d_projection() was deprecated in Matplotlib "
+                        "%(since)s and will be removed %(removal)s.")
```

Comment:
Can you also update the removal version?

Reply:
Sure - is it for removal in 3.6?

Reply:
Yes.  We should have removed it in 3.5, but that change is too big to do in a patch release so we have to delay it.

Reply:
Wait? 3.4 deprecation is scheduled for removal in 3.6. I think you've confused this with a different deprecation.

Reply:
As it stands, the message says it "was deprecated in Matplotlib 3.4 and will be removed two minor versions later". That's already consistent with 3.6, unless you'd rather it's more specific?

Reply:
Nevermind, @QuLogic is right I'm confused  🐑 .